File: tutorials/023_run_mipi_camera_model/run_mipi_camera_model.cpp
// Run a model from a MIPI/libcamera camera source.
//
// Usage:
//   tutorial_023_run_mipi_camera_model --model /path/to/model.tar.gz [--frames 5]

#include <neat.h>

#include <algorithm>
#include <cctype>
#include <cstdint>
#include <filesystem>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>

namespace neat = simaai::neat;
namespace fs = std::filesystem;

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

bool has_flag(int argc, char** argv, const std::string& key) {
  for (int i = 1; i < argc; ++i) {
    if (key == argv[i])
      return true;
  }
  return false;
}

int int_arg(int argc, char** argv, const std::string& key, int def) {
  std::string value;
  if (!get_arg(argc, argv, key, value))
    return def;
  return std::stoi(value);
}

std::string lower_copy(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) {
    return static_cast<char>(std::tolower(c));
  });
  return value;
}

neat::BoxDecodeType decode_type_from_token(const std::string& token) {
  const std::string v = lower_copy(token);
  if (v.empty() || v == "none" || v == "raw")
    return neat::BoxDecodeType::Unspecified;
  if (v == "yolo")
    return neat::BoxDecodeType::Yolo;
  if (v == "yolov5")
    return neat::BoxDecodeType::YoloV5;
  if (v == "yolov8")
    return neat::BoxDecodeType::YoloV8;
  if (v == "yolov8seg" || v == "yolov8-seg")
    return neat::BoxDecodeType::YoloV8Seg;
  if (v == "yolov9")
    return neat::BoxDecodeType::YoloV9;
  if (v == "yolov9seg" || v == "yolov9-seg")
    return neat::BoxDecodeType::YoloV9Seg;
  throw std::runtime_error("unsupported --decode token: " + token);
}

template <typename Shape>
std::string shape_string(const Shape& shape) {
  std::string out = "[";
  for (std::size_t i = 0; i < shape.size(); ++i) {
    out += std::to_string(shape[i]);
    if (i + 1 < shape.size())
      out += ",";
  }
  out += "]";
  return out;
}

neat::Model::Options model_options_for_camera(const neat::CameraInputOptions& camera,
                                              neat::BoxDecodeType decode_type) {
  neat::Model::Options options;
  options.preprocess.kind = neat::InputKind::Image;
  options.preprocess.input_max_width = static_cast<int>(camera.width);
  options.preprocess.input_max_height = static_cast<int>(camera.height);
  options.preprocess.input_max_depth = 3;
  options.preprocess.color_convert.input_format = neat::PreprocessColorFormat::NV12;
  options.preprocess.color_convert.output_format = neat::PreprocessColorFormat::RGB;
  options.preprocess.resize.enable = neat::AutoFlag::On;
  options.preprocess.resize.width = 640;
  options.preprocess.resize.height = 640;
  options.preprocess.resize.mode = neat::ResizeMode::Letterbox;
  options.preprocess.resize.pad_value = 114;
  options.preprocess.preset = neat::NormalizePreset::COCO_YOLO;
  options.advanced_execution.preprocess_target = "EV74";
  options.decode_type = decode_type;
  if (decode_type == neat::BoxDecodeType::Unspecified) {
    options.inference_terminal.mla_only = true;
  } else {
    options.advanced_execution.postprocess_target = "EV74";
    options.score_threshold = 0.25f;
    options.nms_iou_threshold = 0.45f;
    options.top_k = 100;
  }
  return options;
}

void usage(const char* argv0) {
  std::cerr << "Usage: " << argv0
            << " --model <model.tar.gz> [--frames 5] [--width 1920] [--height 1080] "
               "[--fps 30] [--camera-name NAME] [--decode none|yolov8|yolov9seg] "
               "[--pull-timeout-ms 15000] [--strict-zero-copy] [--print-backend]\n";
}

} // namespace

int main(int argc, char** argv) {
  try {
    std::string model_path;
    if (!get_arg(argc, argv, "--model", model_path)) {
      usage(argv[0]);
      return 1;
    }
    if (!fs::exists(model_path))
      throw std::runtime_error("model archive not found: " + model_path);

    const int frames = int_arg(argc, argv, "--frames", 5);
    if (frames <= 0)
      throw std::runtime_error("--frames must be positive");
    const int pull_timeout_ms = int_arg(argc, argv, "--pull-timeout-ms", 15000);
    if (pull_timeout_ms <= 0)
      throw std::runtime_error("--pull-timeout-ms must be positive");

    std::string decode_token = "none";
    get_arg(argc, argv, "--decode", decode_token);

    // CORE LOGIC
    // STEP configure-camera
    neat::CameraInputOptions camera;
    camera.width = static_cast<std::uint32_t>(int_arg(argc, argv, "--width", 1920));
    camera.height = static_cast<std::uint32_t>(int_arg(argc, argv, "--height", 1080));
    camera.framerate_num = static_cast<std::uint32_t>(int_arg(argc, argv, "--fps", 30));
    camera.framerate_den = 1;
    camera.format = "NV12";
    camera.buffer_name = "camera0";
    camera.allow_cpu_fallback = !has_flag(argc, argv, "--strict-zero-copy");
    std::string camera_name;
    if (get_arg(argc, argv, "--camera-name", camera_name)) {
      camera.camera_name = camera_name;
    }
    // END STEP

    // STEP configure-model
    const neat::BoxDecodeType decode_type = decode_type_from_token(decode_token);
    neat::Model model(model_path, model_options_for_camera(camera, decode_type));

    neat::Model::RouteOptions route;
    route.include_input = false;
    route.include_output = true;
    route.upstream_name = camera.buffer_name;
    route.buffer_name = camera.buffer_name;
    route.name_suffix = "_camera0";
    route.advanced_execution.preprocess_target = "EV74";
    if (decode_type != neat::BoxDecodeType::Unspecified) {
      route.advanced_execution.postprocess_target = "EV74";
    }
    // END STEP

    // STEP compose-graph
    neat::Graph graph("mipi_camera_model");
    graph.add(neat::nodes::CameraInput(camera));
    graph.add(model.graph(route));

    if (has_flag(argc, argv, "--print-backend")) {
      std::cout << graph.describe_backend(false) << "\n";
    }

    neat::Run run = graph.build();
    // END STEP

    // STEP pull-output
    for (int i = 0; i < frames; ++i) {
      std::optional<neat::Sample> sample = run.pull(/*timeout_ms=*/pull_timeout_ms);
      if (!sample.has_value()) {
        std::cout << "frame=" << i << " output_timeout timeout_ms=" << pull_timeout_ms;
        const std::string last_error = run.last_error();
        if (!last_error.empty())
          std::cout << " last_error=" << last_error;
        std::cout << "\n";
        return 2;
      }
      const neat::TensorList tensors = neat::tensors_from_sample(*sample, true);
      std::cout << "frame=" << i << " tensors=" << tensors.size();
      if (!tensors.empty())
        std::cout << " first_shape=" << shape_string(tensors.front().shape);
      std::cout << "\n";
    }
    // END STEP
    // END CORE LOGIC

    std::cout << "[OK] 023_run_mipi_camera_model\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}

File: tutorials/023_run_mipi_camera_model/run_mipi_camera_model.py
#!/usr/bin/env python3
"""Run a model from a MIPI/libcamera camera source.

Usage:
  python3 run_mipi_camera_model.py --model /path/to/model.tar.gz [--frames 5]
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )


def decode_type_from_token(token: str):
  v = token.strip().lower().replace("-", "")
  if v in ("", "none", "raw"):
    return pyneat.BoxDecodeType.Unspecified
  if v == "yolo":
    return pyneat.BoxDecodeType.Yolo
  if v == "yolov5":
    return pyneat.BoxDecodeType.YoloV5
  if v == "yolov8":
    return pyneat.BoxDecodeType.YoloV8
  if v == "yolov8seg":
    return pyneat.BoxDecodeType.YoloV8Seg
  if v == "yolov9":
    return pyneat.BoxDecodeType.YoloV9
  if v == "yolov9seg":
    return pyneat.BoxDecodeType.YoloV9Seg
  raise ValueError(f"unsupported --decode token: {token}")


def model_options_for_camera(camera: pyneat.CameraInputOptions, decode_type) -> pyneat.ModelOptions:
  options = pyneat.ModelOptions()
  options.preprocess.kind = pyneat.InputKind.Image
  options.preprocess.input_max_width = int(camera.width)
  options.preprocess.input_max_height = int(camera.height)
  options.preprocess.input_max_depth = 3
  options.preprocess.color_convert.input_format = pyneat.PreprocessColorFormat.NV12
  options.preprocess.color_convert.output_format = pyneat.PreprocessColorFormat.RGB
  options.preprocess.resize.enable = pyneat.AutoFlag.On
  options.preprocess.resize.width = 640
  options.preprocess.resize.height = 640
  options.preprocess.resize.mode = pyneat.ResizeMode.Letterbox
  options.preprocess.resize.pad_value = 114
  options.preprocess.preset = pyneat.NormalizePreset.COCO_YOLO
  options.advanced_execution.preprocess_target = "EV74"
  options.decode_type = decode_type
  if decode_type == pyneat.BoxDecodeType.Unspecified:
    options.inference_terminal.mla_only = True
  else:
    options.advanced_execution.postprocess_target = "EV74"
    options.score_threshold = 0.25
    options.nms_iou_threshold = 0.45
    options.top_k = 100
  return options


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--model", type=Path, required=True)
  ap.add_argument("--frames", type=int, default=5)
  ap.add_argument("--width", type=int, default=1920)
  ap.add_argument("--height", type=int, default=1080)
  ap.add_argument("--fps", type=int, default=30)
  ap.add_argument("--camera-name", default="")
  ap.add_argument("--decode", default="none", help="none, yolov8, yolov8seg, yolov9, ...")
  ap.add_argument("--pull-timeout-ms", type=int, default=15000)
  ap.add_argument("--strict-zero-copy", action="store_true")
  ap.add_argument("--print-backend", action="store_true")
  args = ap.parse_args(argv[1:])

  if not args.model.exists():
    raise FileNotFoundError(f"model archive not found: {args.model}")
  if args.frames <= 0:
    raise ValueError("--frames must be positive")
  if args.pull_timeout_ms <= 0:
    raise ValueError("--pull-timeout-ms must be positive")

  # CORE LOGIC
  # STEP configure-camera
  camera = pyneat.CameraInputOptions()
  camera.width = args.width
  camera.height = args.height
  camera.framerate_num = args.fps
  camera.framerate_den = 1
  camera.format = "NV12"
  camera.buffer_name = "camera0"
  camera.allow_cpu_fallback = not args.strict_zero_copy
  if args.camera_name:
    camera.camera_name = args.camera_name
  # END STEP

  # STEP configure-model
  decode_type = decode_type_from_token(args.decode)
  model = pyneat.Model(str(args.model), model_options_for_camera(camera, decode_type))

  route = pyneat.ModelRouteOptions()
  route.include_input = False
  route.include_output = True
  route.upstream_name = camera.buffer_name
  route.buffer_name = camera.buffer_name
  route.name_suffix = "_camera0"
  route.advanced_execution.preprocess_target = "EV74"
  if decode_type != pyneat.BoxDecodeType.Unspecified:
    route.advanced_execution.postprocess_target = "EV74"
  # END STEP

  # STEP compose-graph
  graph = pyneat.Graph("mipi_camera_model")
  graph.add(pyneat.nodes.camera_input(camera))
  graph.add(model.graph(route))

  if args.print_backend:
    print(graph.describe_backend(False))

  run = graph.build()
  # END STEP

  # STEP pull-output
  for i in range(args.frames):
    sample = run.pull(timeout_ms=args.pull_timeout_ms)
    if sample is None:
      msg = f"frame={i} output_timeout timeout_ms={args.pull_timeout_ms}"
      last_error = run.last_error()
      if last_error:
        msg += f" last_error={last_error}"
      print(msg)
      return 2
    tensors = list(sample.tensors)
    print(f"frame={i} tensors={len(tensors)}", end="")
    if tensors:
      print(f" first_shape={list(tensors[0].shape)}", end="")
    print()
  # END STEP
  # END CORE LOGIC

  print("[OK] 023_run_mipi_camera_model")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))
